compiler/qsc_parse/src/ty.rs

```diff
@@ -72,9 +73,53 @@ pub(super) fn array_or_arrow(s: &mut ParserContext<'_>, mut lhs: Ty, lo: u32) ->
     }
 }
 
-pub(super) fn param(s: &mut ParserContext) -> Result<Box<Ident>> {
+pub(super) fn param(s: &mut ParserContext) -> Result<TypeParameter> {
     throw_away_doc(s);
-    apos_ident(s)
+    let lo = s.peek().span.lo;
+    let generic = apos_ident(s)?;
+    let bounds = if token(s, TokenKind::Colon).is_ok() {
+        Some(class_constraints(s)?)
+    } else {
+        None
+    };
+
+    Ok(TypeParameter::new(
+        *generic,
+        bounds.unwrap_or_else(|| ClassConstraints(Box::new([]))),
+        s.span(lo),
+    ))
+}
+
+/// Parses the bounds of a type parameter, which are a list of class names separated by `+`.
+/// This occurs after a `:` in a generic type:
+/// `T: Eq + Iterator[Bool] + Class3`
+///     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ bounds
+fn class_constraints(s: &mut ParserContext) -> Result<ClassConstraints> {
+    let mut bounds: Vec<ClassConstraint> = Vec::new();
+    loop {
+        s.expect(WordKinds::PrimitiveClass);
+        let bound_name = ident(s)?;
+        // if there's a less-than sign, or "open angle bracket", try to parse type parameters for
+        // the class
+        // e.g. `Iterator[Bool]`
```

Comment:
code is correct, but comment is out of date, should be about angle brackets:

```suggestion
        // if there's an "open bracket", try to parse type parameters for the class
        // e.g. `Iterator[Bool]`
```